Add endpoints to manage images and containers locally

pcarranzav opened the pull request:
connects to #142 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/resin-io/resin-supervisor/143)
<!-- Reviewable:end -->

Page- commented on src/api.coffee:
```diff
@@ -16,7 +17,7 @@ privateAppEnvVars = [
 
 module.exports = (application) ->
 	api = express()
-	api.use(bodyParser())
```
You could do something like:
```coffee

unparsedRouter = express.Router()
api.post '/v1/images/create', dockerUtils.createImage
...
parsedRouter = express.Router()
parsedRouter.use(bodyParser())
parsedRouter.post '/v1/update', ...
...
api.use(parsedRouter)
api.use(parsedRouter)
```
in order to separate routes that require body parsing and routes that don't.  It might even make sense to do something like
```coffee
unparsedRouter = express.Router()
api.post '/images/create', dockerUtils.createImage
app.use('/v1/local/', unparsedRouter)
```
to completely isolate it under its own path

pcarranzav replied:
Very nice, thanks @Page-

pcarranzav replied:
(Done)
Did not use a namespace as endpoints that use parsed bodies are in several different actions, including images and containers.

pcarranzav commented:
Only missing CHANGELOG entries and documentation.
Endpoints have been tested, but thorough testing of the whole API would also be required.

Page- commented on docs/API.md:
```diff
@@ -484,3 +484,129 @@ $ curl -X POST --header "Content-Type:application/json" \
+
+## API endpoints for image and container management
+
+Supervisor 1.9 introduces these endpoins to allow users to create images and containers from the app. These endpoints are designed to be
```
`endpoins` -> `endpoints`

pcarranzav replied:
Ouch, thans

Page- commented on src/docker-utils.coffee:
```diff
@@ -162,51 +169,164 @@ do ->
+	exports.startContainer = (req, res) ->
+		docker.getContainer(req.params.id).startAsync(req.body)
+		.then (data) ->
+				res.json(data)
```
This is doubly indented

Page- commented on src/docker-utils.coffee:
```diff
@@ -162,51 +169,164 @@ do ->
+	exports.stopContainer = (req, res) ->
+		docker.getContainer(req.params.id).stopAsync(sanitizeQuery(req.query))
+		.then (data) ->
+				res.json(data)
```
This is doubly indented

Page- commented on src/docker-utils.coffee:
```diff
@@ -162,51 +169,164 @@ do ->
+	exports.deleteContainer = (req, res) ->
+		docker.getContainer(req.params.id).removeAsync(sanitizeQuery(req.query))
+		.then (data) ->
+				res.json(data)
```
This is doubly indented

pcarranzav replied:
Oops again, fixed these three

pcarranzav commented:
@Page- the thing I'm most unsure about here is the repoTag story, regarding how we build it from the createImage request and how we compare against it to avoid cleaning up the image. And the fact that we require users to use the same repo:tag when creating containers cause we have that saved in the DB.
Any insight from your experience with this kind of stuff? Does it look good to you? Or any ideas on how to improve it?

lekkas commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/API.md, line 515 \[r4\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHtaUu8JtVKODW8a992:-KHtaUu9pBNNTcqhlG0d:-649438841) ([raw file](https://github.com/resin-io/resin-supervisor/blob/373aa2e7a8316dee691c8606c555a4b620b30027/docs/API.md#L515)):*
> ```Markdown
> Pulling from DockerHub:
> ```bash
> $ curl -X POST --data-binary @hello-master.tar \
> ```

Is this required when pulling from DockerHub?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143)*
<!-- Sent from Reviewable.io -->

lekkas commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/API.md, line 605 \[r4\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHtbY7dQdWUHFtZ3Fm3:-KHtbY7e_tuwTZcGT9WS:1956215654) ([raw file](https://github.com/resin-io/resin-supervisor/blob/373aa2e7a8316dee691c8606c555a4b620b30027/docs/API.md#L605)):*
> ```Markdown
> 
> Lists containers. Works like [GET /containers/json from the Docker API](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#list-containers).
> The id can be extracted from the response of /v1/containers/create.
> ```

This can be removed - no id required here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143)*
<!-- Sent from Reviewable.io -->

lekkas commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/docker-utils.coffee, line 313 \[r4\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHtgX8QWV5tR7K6RKQb:-KHtgX8QWV5tR7K6RKQc:1803386733) ([raw file](https://github.com/resin-io/resin-supervisor/blob/373aa2e7a8316dee691c8606c555a4b620b30027/src/docker-utils.coffee#L313)):*
> ```CoffeeScript
> 			res.status(500).send(err?.message or err or 'Unknown error')
> 
> 	exports.stopContainer = (req, res) ->
> ```

Do we allow our users to stop app/supervisor containers? I mean, the `listContainers` endpoint will return all containers including supervisor, right? The same comment applies to `/deleteContainer` too. I guess this isn't really a problem if supervisor is automatically restarted by the HostOS though

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143)*
<!-- Sent from Reviewable.io -->

Page- commented:
SGTM

pcarranzav commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[docs/API.md, line 515 \[r4\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHtaUu8JtVKODW8a992:-KHvlq-j9mTTiaPidorK:1758359711) ([raw file](https://github.com/resin-io/resin-supervisor/blob/373aa2e7a8316dee691c8606c555a4b620b30027/docs/API.md#L515)):*
<details><summary><i>Previously, lekkas (Kostas Lekkas) wrote…</i></summary>
> Is this required when pulling from DockerHub?</details>
Fixed.

---

*[docs/API.md, line 605 \[r4\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHtbY7dQdWUHFtZ3Fm3:-KHvlt0sc1VQDJSKoGH_:1758359711) ([raw file](https://github.com/resin-io/resin-supervisor/blob/373aa2e7a8316dee691c8606c555a4b620b30027/docs/API.md#L605)):*
<details><summary><i>Previously, lekkas (Kostas Lekkas) wrote…</i></summary>
> This can be removed - no id required here.</details>
Fixed.

---

*[src/docker-utils.coffee, line 309 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHVGprmN1jTZbyC-qve-r3-309:-KHvlwrMhgR8eSU_QkuI:1758359711) ([raw file](https://github.com/resin-io/resin-supervisor/blob/18997ee614d3818b385fe34cfeb9aed40b851932/src/docker-utils.coffee#L309)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> This is doubly indented</details>
Fixed.

---

*[src/docker-utils.coffee, line 316 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHVGprmN1jTZbyC-qve-r3-316:-KHvm0MEzi7aAsFP22DH:1758359711) ([raw file](https://github.com/resin-io/resin-supervisor/blob/18997ee614d3818b385fe34cfeb9aed40b851932/src/docker-utils.coffee#L316)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> This is doubly indented</details>
Fixed.

---

*[src/docker-utils.coffee, line 313 \[r4\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHtgX8QWV5tR7K6RKQb:-KHvlyrcxUHfO1jt65gb:2057770724) ([raw file](https://github.com/resin-io/resin-supervisor/blob/373aa2e7a8316dee691c8606c555a4b620b30027/src/docker-utils.coffee#L313)):*
<details><summary><i>Previously, lekkas (Kostas Lekkas) wrote…</i></summary>
> Do we allow our users to stop app/supervisor containers? I mean, the `listContainers` endpoint will return all containers including supervisor, right? The same comment applies to `/deleteContainer` too. I guess this isn't really a problem if supervisor is automatically restarted by the HostOS though</details>
Added some protections against this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143)*
<!-- Sent from Reviewable.io -->

Page- commented on src/docker-utils.coffee:
```diff
@@ -162,51 +169,178 @@ do ->
+		container = docker.getContainer(req.params.id)
+		knex('app').select()
+		.then (apps) ->
+			throw new Error('Cannot stop an app container') if _.any(apps, (app) -> req.params.id == app.containerId)
```
This can be `_.any(apps, containerId: req.params.id)`

Page- commented on src/docker-utils.coffee:
```diff
@@ -162,51 +169,178 @@ do ->
+		container = docker.getContainer(req.params.id)
+		nex('app').select()
+		.then (apps) ->
+			throw new Error('Cannot remove an app container') if _.any(apps, (app) -> req.params.id == app.containerId)
```
This can be `_.any(apps, containerId: req.params.id)`

Page- commented:
SGTM

lekkas commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, Page- wrote…</i></summary>
> SGTM</details>


---

Reviewed 1 of 3 files at r1, 1 of 2 files at r2, 1 of 3 files at r3, 1 of 1 files at r5, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-:-KHy-Vzic0PszoRFpW4x:1417088721)*
<!-- Sent from Reviewable.io -->

pcarranzav commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/docker-utils.coffee, line 331 \[r6\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143#-KHVGprmN1jTZbyC-qve-r6-331:-KHzOQJZmsksw1jVbP4Z:-499146009) ([raw file](https://github.com/resin-io/resin-supervisor/blob/3318b0cef89ba6287ada3d3d1334963ac8b0e88b/src/docker-utils.coffee#L331)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> This can be `_.any(apps, containerId: req.params.id)`</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/143)*
<!-- Sent from Reviewable.io -->
